### PR TITLE
C#: Reuse RequestCallContextSafeHandle objects by pooling them.

### PR DESCRIPTION
**Pull request by jtattermusch**
Similar as https://github.com/grpc/grpc/pull/13476, but this time for RequestCallContext objects.

### CODE REVIEW

**Comment by jtattermusch** [comment]
[ignore this\ trying to kick the cla bot

File: src/csharp/Grpc.Microbenchmarks/SendMessageBenchmark.cs [hunk 1]
```diff
@@ -52,7 +52,7 @@ public void Run(int threadCount, int iterations, int payloadSize)
 
         private void ThreadBody(int iterations, int payloadSize)
         {
-            var completionRegistry = new CompletionRegistry(environment, () => environment.BatchContextPool.Lease());
+            var completionRegistry = new CompletionRegistry(environment, () => environment.BatchContextPool.Lease(), () => RequestCallContextSafeHandle.Create());
```
Comment (apolcyn) [inline src/csharp/Grpc.Microbenchmarks/SendMessageBenchmark.cs]: it looks a little strange how this is using `RequestCallContextSafeHandle.Create()` without the pool, even though it doesn't matter and is unused.

Replace `() => RequestCallContextSafeHandle.Create()` with something like `() => null` for clarity?

File: src/csharp/Grpc.Microbenchmarks/CompletionRegistryBenchmark.cs [hunk 2]
```diff
@@ -43,15 +43,15 @@ public void Cleanup()
         public void Run(int threadCount, int iterations, bool useSharedRegistry)
         {
             Console.WriteLine(string.Format("CompletionRegistryBenchmark: threads={0}, iterations={1}, useSharedRegistry={2}", threadCount, iterations, useSharedRegistry));
-            CompletionRegistry sharedRegistry = useSharedRegistry ? new CompletionRegistry(environment, () => BatchContextSafeHandle.Create()) : null;
+            CompletionRegistry sharedRegistry = useSharedRegistry ? new CompletionRegistry(environment, () => BatchContextSafeHandle.Create(), () => RequestCallContextSafeHandle.Create()) : null;
```
Comment (apolcyn) [inline src/csharp/Grpc.Microbenchmarks/CompletionRegistryBenchmark.cs]: might be forgetting here, is this creating the `CompletionRegistry` registry the old way just for comparison?

Comment (jtattermusch) [reply src/csharp/Grpc.Microbenchmarks/CompletionRegistryBenchmark.cs]: This benchmark actually doesn't use the factories from completion registry at all, it uses a ctx object created here: https://github.com/grpc/grpc/pull/13521/files#diff-ba961c28d30e9e908894c6d111e75d4bR55

Comment (apolcyn) [reply src/csharp/Grpc.Microbenchmarks/CompletionRegistryBenchmark.cs]: I see. 

one remaining comment on the `SendMessageBenchmark.cs`, otherwise lgtm

**Comment by jtattermusch** [comment]
Addressed comment PTAL.

**Pull request merged**